jobs: retry non-cancelable running and all reverting jobs

sajjadrizvi opened the pull request:
Previously, non-cancelable jobs were retried in running state
only if their errors were marked as retryable. Moreover,  only 
non-cancelable reverting jobs were retried by default. This 
commit makes non-cancelable jobs always retry in running 
state unless their error is marked as a permanent error. In
addition, this commit makes all reverting jobs to retry when
they fail. As a result, non-cancelable running jobs and all
reverting jobs do not fail due to transient errors.

Release justification: low-risk updates to new functionality.

Release note (general change): Non-cancelable jobs, such as
schema-change GC jobs, now do not fail unless they fail with
a permanent error. They retry with exponential-backoff if
they fail due to a transient error. Furthermore, Jobs that 
perform reverting tasks do not fail. Instead, they are retried 
with exponential-backoff if an error is encountered while 
reverting. As a result, transient errors do not impact the jobs 
that are reverting.

Fixes: #66685

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69300)
<!-- Reviewable:end -->

sajjadrizvi reviewed (commented):
Please pay attention to the tests that I have modified. I have tried to make minimal changes and not beat the purpose of the tests, but I might have overlooked some cases.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MhxxCU30tpedl6DFlxs:bwqx0ph)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I'm confused by this change. It seems to have changed the behavior from always retrying non-cancelable jobs to always retrying revertible jobs. Those are not the same thing. We want to always retry non-cancelable or reverting jobs. One way to unify them is to say that jobs which are reverting are implicitly non-cancelable. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MhyRVV38aX1zuDBMTng:b-s2ypsy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sajjadrizvi)

___
*[pkg/jobs/jobs_test.go, line 3211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MhyQw3R0X6kbmOKwJVM:-MhyQw3R0X6kbmOKwJVN:b-mi1git) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fb6bd70f108acef7cbfa063f852e3326b1b3be6/pkg/jobs/jobs_test.go#L3211)):*
> ```
> // TestNonCancelableJobsRetry tests that a non-cancelable job is retried when
> // failed with a non-retryable error.
> func TestNonCancelableJobsRetry(t *testing.T) {
> ```
I thought that our previous intention was that non-cancelable jobs will retry unless they fail with an error which is marked as permanent. This first error does not seem to be marked as permanent. Why are we not retrying it? Was that logic only for the `OnFailOrCancel` error? That's a bug. I'm also confused by the renaming here. Isn't the point to exercise `NonCancelable`? If `NonCancelable` only affect retry behavior of `OnFailOrCancel` then it doesn't affect the retry behavior of anything anymore. 
___
*[pkg/jobs/registry.go, line 1294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MhyRRxm5Ew9ZNJNsjw_:-MhyRRxm5Ew9ZNJNsjwa:b-viohr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9388792483bb2e40ed7c4e42237946c247a3ad76/pkg/jobs/registry.go#L1294)):*
> ```Go
> 		}
> 		// A non-cancelable job is always retried while reverting unless the error is marked as permanent.
> 		if job.Payload().Noncancelable && !IsPermanentJobError(err) {
> ```

why is this behavior gone?


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> We want to always retry non-cancelable or reverting jobs.

I think I missed a point here. I thought we want to retry non-cancelable jobs forever only when they are `reverting`. So I have not modified anything for the running state. So, is this what we want: `In running state, retry non-cancelable jobs if their error is not marked as a permanent error. In reverting state, retry all jobs regardless of the error type and their non-cancelable status`. Is this correct?

> One way to unify them is to say that jobs which are reverting are implicitly non-cancelable.

This is what I intended to do in this PR. Instead of always retrying only non-cancelable jobs, we retry all jobs that fail in their reverting state.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MhyZAoxCkWW_yMnFGcM:b-gbirwi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/jobs_test.go, line 3211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MhyQw3R0X6kbmOKwJVM:-MhyaHQ6--pYahk0OW7w:bppr27z) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fb6bd70f108acef7cbfa063f852e3326b1b3be6/pkg/jobs/jobs_test.go#L3211)):*
> I thought that our previous intention was that non-cancelable jobs will retry unless they fail with an error which is marked as permanent.

Does it include `running` state as well? I thought we want this only for reverting state.

> This first error does not seem to be marked as permanent. Why are we not retrying it?

Are you talking about the `running` state? I thought we don't want to modify the behaviour in `running` state. I think it is retried in the `reverting` state. Otherwise, the test should not pass.
___
*[pkg/jobs/registry.go, line 1294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MhyRRxm5Ew9ZNJNsjw_:-MhyaYdz5SJTmIPnjhMp:b-8dyppz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9388792483bb2e40ed7c4e42237946c247a3ad76/pkg/jobs/registry.go#L1294)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why is this behavior gone?
</blockquote></details>

Because we don't have to check error type or non-cancelable status to decide to retry. Don't we want to retry all jobs that are reverting?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> I think I missed a point here. I thought we want to retry non-cancelable jobs forever only when they are reverting. So I have not modified anything for the running state. So, is this what we want: In running state, retry non-cancelable jobs if their error is not marked as a permanent error. In reverting state, retry all jobs regardless of the error type and their non-cancelable status. Is this correct?

Yes, this is correct. 

> This is what I intended to do in this PR. Instead of always retrying only non-cancelable jobs, we retry all jobs that fail in their reverting state.

`Instead of always retrying only non-cancelable jobs, we retry all jobs that fail in their reverting state.` and `One way to unify them is to say that jobs which are reverting are implicitly non-cancelable.` sounds to me like they are in conflict. One says the union of non-cancelable and reverting, the other says, just reverting. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MhylYmQDaGNoMGLf1Os:b7nnx2f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/jobs/jobs_test.go, line 3211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MhyQw3R0X6kbmOKwJVM:-MhylxDj4bXM-y_dzwx5:bpuxcaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fb6bd70f108acef7cbfa063f852e3326b1b3be6/pkg/jobs/jobs_test.go#L3211)):*
<details><summary><i>Previously, sajjadrizvi (Sajjad Rizvi) wrote…</i></summary><blockquote>

> I thought that our previous intention was that non-cancelable jobs will retry unless they fail with an error which is marked as permanent.

Does it include `running` state as well? I thought we want this only for reverting state.

> This first error does not seem to be marked as permanent. Why are we not retrying it?

Are you talking about the `running` state? I thought we don't want to modify the behaviour in `running` state. I think it is retried in the `reverting` state. Otherwise, the test should not pass.
</blockquote></details>

We want to retry all non-cancelable jobs even if they are in the running state. 


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> Yes, this is correct.

OK, let me update it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-Mhyv5Yb6Tt87X52X0sM:b-wjrl9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/jobs_test.go, line 3211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MhyQw3R0X6kbmOKwJVM:-MhyvFew8rWcSdkW7Yw-:b1p8gti) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fb6bd70f108acef7cbfa063f852e3326b1b3be6/pkg/jobs/jobs_test.go#L3211)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

We want to retry all non-cancelable jobs even if they are in the running state. 
</blockquote></details>

I am now adding it.


<!-- Sent from Reviewable.io -->

Commit:
jobs: retry non-cancelable running and all reverting jobs

Previously, non-cancelable jobs were retried in running state
only if their errors were marked as retryable. Moreover, only
non-cancelable reverting jobs were retried by default. This
commit makes non-cancelable jobs always retry in running
state unless their error is marked as a permanent error. In
addition, this commit makes all reverting jobs to retry when
they fail. As a result, non-cancelable running jobs and all
reverting jobs do not fail due to transient errors.

Release justification: low-risk updates to new functionality.

Release note (general change): Non-cancelable jobs, such as
schema-change GC jobs, now do not fail unless they fail with
a permanent error. They retry with exponential-backoff if
they fail due to a transient error. Furthermore, Jobs that
perform reverting tasks do not fail. Instead, they are retried
with exponential-backoff if an error is encountered while
reverting. As a result, transient errors do not impact jobs that
are reverting.

Fixes: #66685

ajwerner reviewed (commented):
> Reviewed 5 of 5 files at r2, 4 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/jobs/registry.go, line 1218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1d7zd--ucKATZeaVh:-Mi1d7ze9RP_ofzwIfpQ:b-cuh1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/84562fa47e268036ddbf50b2e68b49935d7d50a8/pkg/jobs/registry.go#L1218)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> 		// TODO(spaskob): enforce a limit on retries.
> 		if errors.Is(err, retryJobErrorSentinel) {
> 			jm.ResumeRetryError.Inc(1)
> 			return errors.Errorf("job %d: %s: restarting in background", job.ID(), err)
> 		}
> 		if job.Payload().Noncancelable && !IsPermanentJobError(err) {
> ```
</details>

Let's do this in a single conditional:

```
if nonCancelableRetry := job.Payload().Noncancelable && !IsPermanentJobError(err); nonCancelableRetry || 
    errors.Is(err, retryJobErrorSentinel)  {
    if nonCancelableRetry {
        err = errors.Wrap(err, "non-cancelable")
    }
    jm.ResumeRetryError.Inc(1)
    // ...
} 
```

___
*[pkg/jobs/testing_knobs.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1g7FQ-q0cH04lozDO:-Mi1g7FQ-q0cH04lozDP:bh0lh1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/jobs/testing_knobs.go#L73)):*
> ```Go
> 	// ModifyErrorAfterOnFailOrCancel captures the error returned from OnFailOorCancel
> 	// and sets the error to the returned value of this function.
> 	ModifyErrorAfterOnFailOrCancel func(jobspb.JobID, error) error
> ```

I'm skeptical of this knob mostly because all the places it seems to be used it is doing something to an error that was already injected by that test. That has a code smell. Please instead adjust those tests to make sense in light of the behavior change. That might mean that instead of looking for a reverting failed state or something like that that instead they just confirm that retries happen a few times and then just finish. Jobs don't need to all have exited before the test concludes. 
___
*[pkg/sql/schema_changer_test.go, line 6377 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1eVrS6BcGWYJihEXf:-Mi1eVrS6BcGWYJihEXg:b3amfdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/sql/schema_changer_test.go#L6377)):*
> ```Go
> // back a schema change causes the job to fail, and that the appropriate error
> // is displayed in the jobs table.
> func TestPermanentErrorDuringRollback(t *testing.T) {
> ```

It seems like this test should be renamed to something like 
___
*[pkg/sql/schema_changer_test.go, line 6383 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1ffVB8RYTr4iVoq10:-Mi1ffVB8RYTr4iVoq11:b-5q643a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/sql/schema_changer_test.go#L6383)):*
> ```Go
> 
> 	var sqlDB *gosql.DB
> 	validateError := func(jobID jobspb.JobID, err error) error {
> ```

this does not feel well motivated. It seems to change the control flow of the test, which is already responsible for injecting these errors. Layering another level of filtering here when the error came from the test in the first place makes this feel crufty. If you feel like you *must* do this, then please explain it with ample commentary. It is important that you digest what this test was exercising and modify it in light of the behavior change in the jobs layer. 
___
*[pkg/sql/schema_changer_test.go, line 6449 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1f4t34wjHcyrx7mo9:-Mi1f4t34wjHcyrx7moA:b-794qxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/sql/schema_changer_test.go#L6449)):*
> ```Go
> 				},
> 			},
> 			// Decrease the adopt-loop interval so that retries happen quickly.
> ```

I don't have strong feelings here but this feels odd to me. 


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/registry.go, line 1218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1d7zd--ucKATZeaVh:-Mi7wOI-4zSlKl75YUO5:b8bkg29) ([raw file](https://github.com/cockroachdb/cockroach/blob/84562fa47e268036ddbf50b2e68b49935d7d50a8/pkg/jobs/registry.go#L1218)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		// TODO(spaskob): enforce a limit on retries.
> 		if errors.Is(err, retryJobErrorSentinel) {
> 			jm.ResumeRetryError.Inc(1)
> 			return errors.Errorf("job %d: %s: restarting in background", job.ID(), err)
> 		}
> 		if job.Payload().Noncancelable && !IsPermanentJobError(err) {
> ```
Let's do this in a single conditional:

```
if nonCancelableRetry := job.Payload().Noncancelable && !IsPermanentJobError(err); nonCancelableRetry || 
    errors.Is(err, retryJobErrorSentinel)  {
    if nonCancelableRetry {
        err = errors.Wrap(err, "non-cancelable")
    }
    jm.ResumeRetryError.Inc(1)
    // ...
} 
</blockquote></details>

I have updated the code.
___
*[pkg/jobs/testing_knobs.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1g7FQ-q0cH04lozDO:-Mi7wVzb4_KjpuWkZPgG:bz6pk6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/jobs/testing_knobs.go#L73)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm skeptical of this knob mostly because all the places it seems to be used it is doing something to an error that was already injected by that test. That has a code smell. Please instead adjust those tests to make sense in light of the behavior change. That might mean that instead of looking for a reverting failed state or something like that that instead they just confirm that retries happen a few times and then just finish. Jobs don't need to all have exited before the test concludes. 
</blockquote></details>

Thanks for the suggestion. My intention was to use the knob only to terminate the test and not retry jobs completely.
___
*[pkg/sql/schema_changer_test.go, line 6377 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1eVrS6BcGWYJihEXf:-Mi7wrdJ9s-eUH8mAKlS:b-ij0b03) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/sql/schema_changer_test.go#L6377)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It seems like this test should be renamed to something like 
</blockquote></details>

Like?? How about `TestRevertingWhileRollingBack`?
___
*[pkg/sql/schema_changer_test.go, line 6383 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1ffVB8RYTr4iVoq10:-Mi7vsya8Bn9ahNzI7-4:bstpbs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/sql/schema_changer_test.go#L6383)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this does not feel well motivated. It seems to change the control flow of the test, which is already responsible for injecting these errors. Layering another level of filtering here when the error came from the test in the first place makes this feel crufty. If you feel like you *must* do this, then please explain it with ample commentary. It is important that you digest what this test was exercising and modify it in light of the behavior change in the jobs layer. 
</blockquote></details>

Thanks for pointing it out. I have removed the testing knob and related code.


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/testing_knobs.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1g7FQ-q0cH04lozDO:-Mi7y4pQCoKQ7i9b6tJx:bd27d7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/jobs/testing_knobs.go#L73)):*
<details><summary><i>Previously, sajjadrizvi (Sajjad Rizvi) wrote…</i></summary><blockquote>

Thanks for the suggestion. My intention was to use the knob only to terminate the test and not retry jobs completely.
</blockquote></details>

I mean, not retry jobs forever.


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
`TestRegistryLifecycle/dump_traces_on_cancel` is getting failed in stress tests because multiple resumers enter `OnFailOrCancel`. It will be fixed once the PR solving this problem is merged in the main branch.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MiDn1G59KP1q1M0WISa:b-chp2vk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> TestRegistryLifecycle/dump_traces_on_cancel is getting failed in stress tests because multiple resumers enter OnFailOrCancel. It will be fixed once the PR solving this problem is merged in the main branch.

Actually not, there is some other problem.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MiDobH_21HXOHrMlpBm:b-35c9w3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/jobs/jobs_test.go, line 315 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MiT8qyo73pLbvcU0Hdz:-MiT8qyo73pLbvcU0He-:bz15ssd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8dac3add77e969c5d5340743a53333bbc7662de/pkg/jobs/jobs_test.go#L315)):*
> ```Go
> 			// TODO(sajjad): To discuss: This success function is not called from Registry
> 			// when the resumer completes without error. Instead, it is being called from
> 			// Resume itself. Is this on purpose or is it a bug?
> ```

I think you could call it a bug. It has to do with the fact that there used to be a "best effort" Success hook but it was fundamentally a broken idea. When somebody refactored that hook away, they didn't want to delete these tests. Their approach was to call this function when Resume returns with no error. 


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/jobs_test.go, line 315 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MiT8qyo73pLbvcU0Hdz:-MibI4bbC-4F7mkMMPE1:b-g3g0li) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8dac3add77e969c5d5340743a53333bbc7662de/pkg/jobs/jobs_test.go#L315)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think you could call it a bug. It has to do with the fact that there used to be a "best effort" Success hook but it was fundamentally a broken idea. When somebody refactored that hook away, they didn't want to delete these tests. Their approach was to call this function when Resume returns with no error. 
</blockquote></details>

OK, I will file an issue to remove this. This is a little confusing. 


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
@ajwerner PHAL! Thanks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-Mic51bw5EpP2P5YTqIg:bglqse6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

Commit:
sql: make job cancelable when adding a table without mutations

This is a small fix to make the jobs cancelable that are
created when we add a table without mutations, e.g., in
`CREATE TABLE foo (x PRIMARY KEY) AS VALUES (1)`.

Release justification: bug fixes and low-risk updates

Release note: None

ajwerner reviewed (commented):
This is quite close. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MjGU6Eb1AZY-5Ohq-_G:b-4ujwzd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/jobs/jobs_test.go, line 804 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGKQVLCbrAeJ8-rTBr:-MjGKQVLCbrAeJ8-rTBs:b-u32t0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/jobs_test.go#L804)):*
> ```Go
> 
> 		// The job will stay in reverting state unless it succeeds reverting. So we don't have
> 		// to exit clean.
> ```

uber-nit: `cleanly`.

Also, why not just have it successfully revert and then move to failed as a clean ending of the test?
___
*[pkg/jobs/jobs_test.go, line 893 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGKcqN6WVT3dLGf032:-MjGKcqN6WVT3dLGf033:b-7vwi6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/jobs_test.go#L893)):*
> ```Go
> 		rts.check(t, jobs.StatusReverting)
> 		require.True(t, triedToMarkFailed.Load().(bool))
> 		// The job is now in state reverting and will never resume again.
> ```

I don't think repeating this comment is helpful. 
___
*[pkg/jobs/jobs_test.go, line 897 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGSbrb2hM9hezXBWWs:-MjGSbrb2hM9hezXBWWt:bh7kr78) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/jobs_test.go#L897)):*
> ```Go
> 		rts.failOrCancelCh <- nil
> 		rts.mu.e.OnFailOrCancelExit = true
> 		// The job will be retried again, However, we don't have to continue the test.
> ```

why not just let the thing succeed in reverting here after one failure?
___
*[pkg/jobs/testing_knobs.go, line 73 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGT4kL8rJZzKUeWTGq:-MjGT4kL8rJZzKUeWTGr:b-sg6nsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/testing_knobs.go#L73)):*
> ```Go
> 	// ModifyErrorAfterOnFailOrCancel captures the error returned from OnFailOorCancel
> 	// and sets the error to the returned value of this function.
> 	ModifyErrorAfterOnFailOrCancel func(jobspb.JobID, error) error
> ```

is this being used anymore?
___
*[pkg/sql/schema_changer_test.go, line 6377 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1eVrS6BcGWYJihEXf:-MjGUAkl2_8CDYIlGCX1:bnkmmfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/sql/schema_changer_test.go#L6377)):*
<details><summary><i>Previously, sajjadrizvi (Sajjad Rizvi) wrote…</i></summary><blockquote>

Like?? How about `TestRevertingWhileRollingBack`?
</blockquote></details>

`TestRetryOnAllErrorsWhenReverting`?
___
*[pkg/sql/schema_changer_test.go, line 6979 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGTbyQ50IXN8DMNDzG:-MjGTbyQ50IXN8DMNDzH:blul4i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/sql/schema_changer_test.go#L6979)):*
> ```Go
> 
> 				go func() {
> 					_, _ = db.Exec(tc.scStmt)
> ```

note that this will not return until server shutdown
___
*[pkg/sql/table.go, line 156 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGTrGeBha7IHXNw2YV:-MjGTrGeBha7IHXNw2YW:bvuxc9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/sql/table.go#L156)):*
> ```Go
> 			// The job should be cancelable when we are adding a table that doesn't
> 			// have mutations, e.g., in CREATE TABLE AS VALUES.
> 			NonCancelable: mutationID == descpb.InvalidMutationID && !tableDesc.Adding(),
> ```

this change is important and deserves its own test. That test should create a `CREATE TABLE ... AS` job and validate that it is not marked as `NonCancelable` despite having an `InvalidMutationID`. You can leverage a hook to do this. Ideally this change and test will be a separate commit. 
___
*[pkg/sql/type_change_test.go, line 228 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGIiki1gNe-3r1JW56:-MjGIikj2tJKn5_YbInC:bkxs4j3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/sql/type_change_test.go#L228)):*
> ```
> 	//tdb := sqlutils.MakeSQLRunner(sqlDB)
> 	//tdb.CheckQueryResultsRetry(t)
> ```
what is this about?


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/jobs_test.go, line 804 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGKQVLCbrAeJ8-rTBr:-MjKHiyzEjF-ho4VUO_z:b-c37grj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/jobs_test.go#L804)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

uber-nit: `cleanly`.

Also, why not just have it successfully revert and then move to failed as a clean ending of the test?
</blockquote></details>

Yeah, I should do that. I have updated accordingly.
___
*[pkg/jobs/jobs_test.go, line 893 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGKcqN6WVT3dLGf032:-MjKHpHs2f3SAixEOrqk:bi79dd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/jobs_test.go#L893)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think repeating this comment is helpful. 
</blockquote></details>

OK, I have removed the comment.
___
*[pkg/jobs/jobs_test.go, line 897 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGSbrb2hM9hezXBWWs:-MjKMV814imUrvNxj2jM:b-22v40i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/jobs_test.go#L897)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why not just let the thing succeed in reverting here after one failure?
</blockquote></details>

yeah, it should be like that! I have updated the test accordingly.
___
*[pkg/jobs/testing_knobs.go, line 73 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGT4kL8rJZzKUeWTGq:-MjKMgr9EUtUZOBfJRcr:b-ij1nax) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/jobs/testing_knobs.go#L73)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

is this being used anymore?
</blockquote></details>

No, I have now removed this knob.
___
*[pkg/sql/schema_changer_test.go, line 6377 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-Mi1eVrS6BcGWYJihEXf:-MjKO-rA1uHeP6Zldu3O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac1a9b92691ae944863a9224c715a0ea4fc67580/pkg/sql/schema_changer_test.go#L6377)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`TestRetryOnAllErrorsWhenReverting`?
</blockquote></details>

Done.
___
*[pkg/sql/table.go, line 156 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjGTrGeBha7IHXNw2YV:-MjQocct9-OPtcUNq-yD:b9qsth7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad89af0152420e3b9742cd3b0242ee1db220f507/pkg/sql/table.go#L156)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this change is important and deserves its own test. That test should create a `CREATE TABLE ... AS` job and validate that it is not marked as `NonCancelable` despite having an `InvalidMutationID`. You can leverage a hook to do this. Ideally this change and test will be a separate commit. 
</blockquote></details>

I have added a test. The change is already in another commit.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 10 files at r9, 4 of 8 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MjUXU104sh8LFfBqhvQ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/sql/schema_changer_test.go, line 2154 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjUXLhC4MitTxbYjZvy:-MjUXLhC4MitTxbYjZvz:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e4779b8c3d7f5ed4d241d05275166670806ed2/pkg/sql/schema_changer_test.go#L2154)):*
> ```Go
> 	const chunkSize = 200
> 	attempts := 0
> 	//syncChan := make(chan struct{})
> ```

?


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
Thanks much for the reviews. I appreciate your thorough review and feedback.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-:-MjU_PFdE78KvI-DcwKh:b60rp5k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schema_changer_test.go, line 2154 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69300#-MjUXLhC4MitTxbYjZvy:-MjUYFIUA8LLU_59VCI9:b-c919bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e4779b8c3d7f5ed4d241d05275166670806ed2/pkg/sql/schema_changer_test.go#L2154)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

?
</blockquote></details>

Sorry, I forgot to remove it.


<!-- Sent from Reviewable.io -->

sajjadrizvi commented:
TFTR!

bors r=ajwerner
